code/drasil-lang/lib/Language/Drasil/Expr/Lang.hs

```diff
@@ -211,10 +213,14 @@ instance LiteralC Expr where
   perc l r = Lit $ perc l r
 
 assocArithOperToTy :: AssocArithOper -> Space
-assocArithOperToTy AddI  = S.Integer
-assocArithOperToTy MulI  = S.Integer
-assocArithOperToTy AddRe = S.Real
-assocArithOperToTy MulRe = S.Real
+assocArithOperToTy Add  = S.Integer
+assocArithOperToTy Add = S.Real
+assocArithOperToTy Add = S.Natural
+assocArithOperToTy Add = S.Rational
+assocArithOperToTy Mul  = S.Integer
+assocArithOperToTy Mul = S.Real
+assocArithOperToTy Mul = S.Natural
+assocArithOperToTy Mul = S.Rational
```

Comment:
You have overlapping pattern matches here, which we can't do. For this area of code (the type-checker), for now, it is enough to ensure that both LHS and RHS of an Add or Mull are of the same (numeric) type, rather than to base what to look for off the "addI" or "mulRe" as it was previously done.

Reply:
How do I ensure that both LHS and RHS of an Add or Mull are of the same (numeric) type?

Reply:
There are functions lying around the definition of `Space` that should help -- I think there's one that determines if a given `Space` is a numeric type or not. From there, I think we can remove this (above, annotated) function entirely, actually.

Reply:
If I delete assocArithOperToTy I need to change AssocA
https://github.com/JacquesCarette/Drasil/blob/85cf74daaccb74131373df9d52cc2bc751fd48e9/code/drasil-lang/lib/Language/Drasil/Expr/Lang.hs#L239-L242)

Do you think changing it to only take two arguments rather then a list of expressions will cause a lot of problems?
  infer cxt (AssocA _ l r) = case (infer cxt l, infer cxt r) of
    (Left lt, Left rt) -> if S.isBasicNumSpace lt && lt == rt
      then Left lt
      else Right $ "Associative arithmetic operation expects all operands to be of the same expected type (" ++ show lt ++ ")." 
    (_, Right x) -> Right x
    (Right x, _) -> Right x

Reply:
What is `strictHead` and `strictLast`? I think `AssocA _ _` carries a list of operands (my bad regarding earlier, I implied it was purely binary!), so we should make sure all items in the list are of the same type.

Reply:
I was not sure how to compare every item in the list so I just compared the first and the last item and if they were both the same number space.
I want to be able to check if each item in the list is of the same space and then get that space. I am looking into how to do that. 
This is how it currently works
https://github.com/JacquesCarette/Drasil/blob/85cf74daaccb74131373df9d52cc2bc751fd48e9/code/drasil-lang/lib/Language/Drasil/Expr/Lang.hs#L239-L242

Reply:
Do we really need `allOfType'` if we already have `allOfType`?

Reply:
Not really